cloudinit/sources/helpers/azure.py

```diff
@@ -1055,4 +1055,213 @@ def dhcp_log_cb(out, err):
     )
 
 
+class BrokenAzureDataSource(Exception):
+    pass
+
+
+class NonAzureDataSource(Exception):
+    pass
+
+
+class OvfEnvXml:
+    OVF_NAMESPACE = "http://schemas.dmtf.org/ovf/environment/1"
+    WA_NAMESPACE = "http://schemas.microsoft.com/windowsazure"
+
+    ns = {
+        "ovf": OVF_NAMESPACE,
+        "wa": WA_NAMESPACE,
+    }
+
+    def __init__(
+        self,
+        *,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        hostname: Optional[str] = None,
+        custom_data: Optional[bytes] = None,
+        disable_ssh_password_auth: Optional[bool] = None,
+        public_keys: Optional[list] = None,
+        preprovisioned_vm: bool = False,
+        preprovisioned_vm_type: Optional[str] = None,
+    ) -> None:
+        self.username = username
+        self.password = password
+        self.hostname = hostname
+        self.custom_data = custom_data
+        self.disable_ssh_password_auth = disable_ssh_password_auth
+        self.public_keys: list = public_keys or []
+        self.preprovisioned_vm = preprovisioned_vm
+        self.preprovisioned_vm_type = preprovisioned_vm_type
+
+    def __eq__(self, other) -> bool:
+        return self.__dict__ == other.__dict__
+
+    def __repr__(self) -> str:
+        props = [
+            "username=%s" % self.username,
+            "password=%s" % self.password,
+            "hostname=%s" % self.hostname,
+            "custom_data=%r" % self.custom_data,
+            "disable_ssh_password_auth=%s" % self.disable_ssh_password_auth,
+            "public_keys=%s" % self.public_keys,
+            "preprovisioned_vm=%s" % self.preprovisioned_vm,
+            "preprovisioned_vm_type=%s" % self.preprovisioned_vm_type,
+        ]
+        return "OvfEnvXml(" + " ".join(props) + ")"
+
+    @classmethod
+    def parse_text(cls, ovf_env_xml: str) -> "OvfEnvXml":
```

Comment:
Is there an advantage?

TIL about `Self`

Reply:
I believe the only advantage would be that it's more semantic for pythons language server recommendations/validations, but it could have no benefit if the language server is already capable of picking up the type on the way you've done it!

Regardless, I wouldn't really classify this as a *blocking* change if you wanted to keep it how it is.